hbase-rest/src/main/java/org/apache/hadoop/hbase/rest/ProtobufHttpEntity.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.rest;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.protocol.HTTP;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * For sending a Protobuf encoded object via Apache HttpClient efficiently without an interim byte
+ * array. This exposes the underlying Apache HttpClient types, but so do the other client classes.
+ */
+@InterfaceAudience.Public
```

Comment:
Does this need to be IA.Public?

Reply:
Yes, we expect the user to explicitly create these entities.

i have not updated the tests to use them, but the intentended usage is similar to how 
org.apache.hadoop.hbase.rest.client.RemoteHTable.put(Put) is implemented.

Reply:
At least have a section in the javadoc to show how to use this class?

And since we have not reach an agreement on whether to move RemoteAdmin and RemoteHTable to src/main, I'm not sure whether we should make this class IA.Public...

Reply:
I only mentined RemoHTable as example where this is used, but this 
is not limited to RemotHTable, it should also be used directly in user code.

I will add a standalone test case that uses this class as an illustration.

Reply:
> At least have a section in the javadoc to show how to use this class?
> 
> And since we have not reach an agreement on whether to move RemoteAdmin and RemoteHTable to src/main, I'm not sure whether we should make this class IA.Public...

This is not tied to RemoteAdmin / RemoteHTable. 
This is just a more efficient way to use the Client class (or a custom client based on Apache HttpClient)

The RemoteAdmin / RemoteHTable  API does not even change, this new class only used in the RemoteHTable implementation.